File: practices/chapter15_greedy/python/ch_15_greedy_pick_radio.py
def greedy_pick_radio(request_cities=set(), radios={}):
  remain_cities = request_cities
  greedy_radios = set()

  while remain_cities:
    greedy_choose_radio = None
    cities_covered_now = set()
    for radio, areas in radios.items():
      # 取交集
      cover_cities = remain_cities & areas
      # 取能覆蓋最大範圍的radio
      if len(cover_cities) > len(cities_covered_now):
        greedy_choose_radio = radio
        cities_covered_now = cover_cities
    # 排除掉覆蓋的cities
    remain_cities -= cities_covered_now
    greedy_radios.add(greedy_choose_radio)
  
  return greedy_radios

request_cities = set([
  'Taipei', 'Keelung', 'Taoyuang', 'Hsinchu',
  'Taichung', 'Chiayi', 'Tainan', 'Kaosiung'
])

all_radios = {
  'Radio 1': set(['Hsinchu', 'Taichung', 'Chiayi']),
  'Radio 2': set(['Keelung', 'Hsinchu', 'Taipei']),
  'Radio 3': set(['Taoyuang', 'Taichung', 'Tainan']),
  'Radio 4': set(['Taichung', 'Chiayi']),
  'Radio 5': set(['Tainan', 'Kaosiung']),
}

def main():
  res = greedy_pick_radio(request_cities, all_radios)
  print(res)

if __name__ == "__main__":
    main()

File: practices/chapter15_greedy/js/ch_15_greedy_pick_radio.js
function greedyPickRadio(requestCities=[], radios={}) {
  let remainCities = [...requestCities]
  let greedyRadios = []

  while (remainCities.length > 0) {
    let chosenRadio = undefined
    let coveredCities = []
    
    for (const radioKey in radios) {
      const radioCities = radios[radioKey]
      // 取交集
      const intersectionCovered = remainCities.filter(
        c => radioCities.includes(c)
      )
      if(intersectionCovered.length > coveredCities.length) {
        chosenRadio = radioKey
        coveredCities = intersectionCovered
      }
    }
    // 取差集
    remainCities = remainCities.filter(
      c => !coveredCities.includes(c)
    )
    greedyRadios.push(chosenRadio)
  }

  return greedyRadios
}

function main() {
  const requestCities = [
    'Taipei', 'Keelung', 'Taoyuang', 'Hsinchu',
    'Taichung', 'Chiayi', 'Tainan', 'Kaosiung'
  ]

  const allRadios = {
    'Radio 1': ['Hsinchu', 'Taichung', 'Chiayi'],
    'Radio 2': ['Keelung', 'Hsinchu', 'Taipei'],
    'Radio 3': ['Taoyuang', 'Taichung', 'Tainan'],
    'Radio 4': ['Taichung', 'Chiayi'],
    'Radio 5': ['Tainan', 'Kaosiung'],
  }

  const res = greedyPickRadio(requestCities, allRadios)
  console.log(res)
}

main()